protocols/autonatv2/src/server/handler/dial_back.rs

```diff
@@ -0,0 +1,148 @@
+            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                result
+                    .map_err(|timeout| io::Error::new(io::ErrorKind::TimedOut, timeout))
+                    .and_then(identity),
+            ));
+        }
+        if let Some(cmd) = self.pending_nonce.pop_front() {
+            self.requested_substream_nonce.push_back(cmd);
+            return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest {
+                protocol: SubstreamProtocol::new(DIAL_BACK_UPGRADE, ()),
+            });
+        }
+        Poll::Pending
+    }
+
+    fn on_behaviour_event(&mut self, event: Self::FromBehaviour) {
+        self.pending_nonce.push_back(event);
+    }
+
+    fn on_connection_event(
+        &mut self,
+        event: ConnectionEvent<
+            Self::InboundProtocol,
+            Self::OutboundProtocol,
+            Self::InboundOpenInfo,
+            Self::OutboundOpenInfo,
+        >,
+    ) {
+        match event {
+            ConnectionEvent::FullyNegotiatedOutbound(FullyNegotiatedOutbound {
+                protocol, ..
+            }) => {
+                if let Some(cmd) = self.requested_substream_nonce.pop_front() {
+                    if self
+                        .outbound
+                        .try_push(perform_dial_back(protocol, cmd))
+                        .is_err()
+                    {
+                        tracing::warn!("Dial back dropped, too many requests in flight");
+                    }
+                } else {
+                    tracing::warn!("received dial back substream without nonce");
+                }
+            }
+            ConnectionEvent::DialUpgradeError(DialUpgradeError {
+                error: StreamUpgradeError::NegotiationFailed,
+                ..
+            })
+            | ConnectionEvent::DialUpgradeError(DialUpgradeError {
+                error: StreamUpgradeError::Timeout,
+                ..
+            }) => {
+                if let Some(cmd) = self.requested_substream_nonce.pop_front() {
+                    let _ = cmd.back_channel.send(DialBackRes::DialBack);
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+async fn perform_dial_back(
+    mut stream: impl AsyncWrite + Unpin,
+    DialBackCommand {
+        nonce,
+        back_channel,
+        ..
+    }: DialBackCommand,
+) -> io::Result<()> {
+    let res = perform_dial_back_inner(&mut stream, nonce)
+        .await
+        .map_err(|_| DialBackRes::DialBack)
+        .map(|_| DialBackRes::Ok)
+        .unwrap_or_else(|e| e);
+    back_channel
+        .send(res)
+        .map_err(|_| io::Error::new(io::ErrorKind::Other, "send error"))?;
+    Ok(())
+}
+
+async fn perform_dial_back_inner(
+    mut stream: impl AsyncWrite + Unpin,
+    nonce: Nonce,
+) -> io::Result<()> {
+    let dial_back = DialBack { nonce };
+    dial_back.write_into(&mut stream).await?;
+    stream.close().await?;
```

Comment:
Instead of exposing `DialBack`, would it make sense to just expose a function from the `request_response` module that takes a nonce and a stream and does the writing + closes it? In other words, the `DialBack` and its `write_into` function don't seem to offer the API we want in this module given that we have to spend an extra 8 lines here to get the functionality that we want.

Reply:
I changed that in my newest commit (soon to be pushed) but that essentially just changes the locations of those lines.